niceman/resource/docker_container.py

```diff
@@ -177,31 +185,47 @@ def _execute_command(self, command, env=None, cwd=None):
         # The following call may throw the following exception:
         #    docker.errors.APIError - If the server returns an error.
         lgr.debug('Running command %r', command)
+        out = []
         execute = self.client.exec_create(container=self.container, cmd=command)
         for i, line in enumerate(
                 self.client.exec_start(exec_id=execute['Id'],
                 stream=True)
         ):
-            if line.startswith(b'rpc error'):
-                raise CommandError(cmd=command, msg="Docker error - %s" % line)
+            out.append(line.rstrip())
             lgr.debug("exec#%i: %s", i, line.rstrip())
 
+        exit_code = self.client.exec_inspect(execute['Id'])['ExitCode']
+        if exit_code not in [0, None]:
+            raise CommandError(cmd=' '.join(command), msg='Exit code: {}'.format(exit_code))
+
+        return (out, None)
+
     # XXX should we start/stop on open/close or just assume that it is running already?
 
 
-    def put(self, src_path, dest_path, preserve_perms=False,
-                owner=None, group=None, recursive=False):
+    def put(self, src_path, dest_path, owner=None, group=None):
         """Take file on the local file system and copy over into the session
         """
-        # self.ssh.put([src_path], remotepath=dest_path)
-        pass
+        tar_stream = io.BytesIO()
+        tar_file = tarfile.TarFile(fileobj=tar_stream, mode='w')
```

Comment:
It is recommended to go through the tar to upload one or more files. There is a "Docker cp" equivalent but it has been depreciated since version 1.2

http://docker-py.readthedocs.io/en/1.5.0/api/#copy

I have added an explanatory comment in the code.